content/blog/scheduling-in-react/index.md

```diff
@@ -0,0 +1,243 @@
+
+## Browser Event Loop
+
+Before we learn more about how this can be achieved, let’s dig deeper and understand why the browser has issues with these kind of user interactions.
+
+JavaScript code is executed in one thread, meaning that only one line of JavaScript can be run at any given time. The same thread is also responsible for other document lifecycles like layout and paint.[^1] This means that whenever JavaScript code runs, the browser is blocked from doing anything else.
+
+To keep the user interface responsive, we only have a very short timeframe before we need to be able to receive the next input events. In the browser run loop visualization presented in Subhie Panicker’s and Jason Miller’s talk, [A Quest to Guarantee Responsiveness](https://developer.chrome.com/devsummit/schedule/scheduling-on-off-main-thread), at the Chrome Dev Summit 2018 below, we can see that we only have 16 milliseconds (on a typical 60Hz screen) before the next frame is drawn and the next events need to be processed:
+
+![The browser event loop starts by running input handlers, followed by animation frame callbacks, and ends with document lifecycles (style, layout, paint). All of this should complete within one frame which is approximately 16 milliseconds on a 60Hz display.](event-loop-browser.png)
+
+Most JavaScript frameworks (including the current version of React) will run updates synchronously. We can think of this as a function `render()` which will only return once the DOM was updated. During this time, the main thread is blocked.
+
+## Problems with Current Solutions
+
+With the information above, we can formulate two problems that we have to solve in order to get to more responsive user interfaces:
+
+1. **Long-running tasks cause frame drops.** We need to make sure all of our tasks are small and can be completed within a couple of milliseconds so that we can run them within one frame.
+
+2. **Different tasks have different priorities.** In the example application above, we saw that prioritizing the user input leads to a better experience overall. To do this, we need a way to define the order and to schedule tasks accordingly.
+
+## Concurrent React and the Scheduler
+
+_⚠️ Warning: The following APIs are not yet stable and will change. I will do my best to keep this post updated._
+
+To implement a properly scheduled user interface with React, we have to look into two new React features:
+
+- **Concurrent React (also known as Time Slicing).** With the help of the new [Fiber architecture](https://www.youtube.com/watch?v=ZCuYPiUIONs) rewrite that was released with React 16, React can now pause during rendering and yield[^2] to the main thread.
+
+  ➡️ With this feature, we’re able to split long-running rendering tasks into small chunks.
+
+- **Scheduler.** A general purpose cooperative main thread scheduler developed by the React Core team that makes it possible to register callbacks with different priority levels in the browser.
```

Comment:
What is in the browser? The registering of callbacks or the priority levels?

"that makes it possible to register callbacks with different priority levels in the browser"

Reply:
The registered callbacks. The priority levels are a concept that the Scheduler is using to communicate the implementation.